_posts/2024-01-25-websockets-in-http4s.md

```diff
@@ -0,0 +1,1561 @@
+  def forUser(targetUser: User): Boolean = targetUser == user
+}
+
+case class ChatMsg(from: User, to: User, msg: String) extends OutputMessage {
+  def forUser(targetUser: User): Boolean = targetUser == to
+}
+```
+
+From the top, the `UnsupportedCommand` will be used when unsupported commands are received, `KeepAlive` will represent `WebSocketFrame.Ping()`, `DiscardMessage` will be used to filter our unwanted message before we send to all users, `SendToUser` will represent messages meant for a particular user and lastly, `ChatMsg` will represent regular chat messages sent by registered users.
+
+### 6.3. Commands
+Next, we'll need to process commands sent through the WebSocket, create a file called `Command.scala`, and add the following code:
+
+```scala
+package rockthejvm.websockets
+
+import cats.Monad
+
+import cats.data.Validated.Valid
+import cats.data.Validated.Invalid
+import cats.syntax.all.*
+
+trait Command[F[_]] {
+  def register(name: String): F[OutputMessage]
+  def isUsernameInUse(name: String): F[Boolean]
+  def enterRoom(user: User, room: Room): F[Seq[OutputMessage]]
+  def chat(user: User, text: String): F[Seq[OutputMessage]]
+  def help(user: User): F[Seq[OutputMessage]]
+  def listRooms(user: User): F[Seq[OutputMessage]]
+  def listMembers(user: User): F[Seq[OutputMessage]]
+  def disconnect(userRef: Ref[F, Option[User]]): F[Seq[OutputMessage]]
+}
+
+object Command {
+  def make[F[_]: Monad](chatState: Ref[F, ChatState]): Command[F] =
+    new Command[F] {
+      override def register(name: String): F[OutputMessage] =
+        User(user) match
+          case Valid(u) =>
+            SuccessfulRegistration(u).pure[F]
+          case Invalid(e) =>
+            ParsingError(None, e.toString).pure[F]
+      override def isUsernameInUse(name: String): F[Boolean] = ???
+      override def enterRoom(user: User, room: Room): F[Seq[OutputMessage]] = ???
+      override def chat(user: User, text: String): F[Seq[OutputMessage]] = ???
+      override def help(user: User): F[Seq[OutputMessage]] = ???      
+      override def listRooms(user: User): F[Seq[OutputMessage]] = ???
+      override def listMembers(user: User): F[Seq[OutputMessage]] = ???
+      override def disconnect(userRef: Ref[F, Option[User]]): F[Seq[OutputMessage]] = ???
+    }
+}
+```
+
+Here we create a `Command` trait with multiple methods, key among which is `register()` that takes a `String` that returns `F[OutputMessage]`, in the companion object we define a `make()` function that takes a `Ref[F, ChatState]` as a parameter and returns a `Command[F]`. 
+
+In the body of the `register()` function we pass `name` to `User` and pattern match against it. If the result is of type `Valid`, we return `SuccessfulRegistration` otherwise we return `ParsingError`.
+
+In both cases, we lift the value into a `F` using the `pure[F]` method which works because we bound `Monad` to `F[_]` in the function definition of `make()`:
```

Comment:
I think it's more correct to say that we bounded F to Monad, rather than vice versa 😄

Reply:
Fixed